src/UriSetPath.c

```diff
@@ -0,0 +1,493 @@
+	 * from cases like "path1:/path2" and "//path1/path2" inside SetPath.
+	 */
+	while (first < afterLast) {
+		switch (first[0]) {
+			case URI_SET_UNRESERVED:
+				break;
+
+			/* pct-encoded */
+			case _UT('%'):
+				if (afterLast - first < 3) {
+					return URI_FALSE;
+				}
+				switch (first[1]) {
+					case URI_SET_HEXDIG:
+						break;
+					default:
+						return URI_FALSE;
+				}
+				switch (first[2]) {
+					case URI_SET_HEXDIG:
+						break;
+					default:
+						return URI_FALSE;
+				}
+				first += 2;
+				break;
+
+			case URI_SET_SUB_DELIMS:
+				break;
+
+			/* ":" / "@" and "/" */
+			case _UT(':'):
+			case _UT('@'):
+			case _UT('/'):
+				break;
+
+			default:
+				return URI_FALSE;
+		}
+
+		first++;
+	}
+	return URI_TRUE;
+}
+
+
+
+static void URI_FUNC(DropEmptyFirstPathSegment)(URI_TYPE(Uri) * uri, UriMemoryManager * memory) {
+	assert(uri != NULL);
+	assert(memory != NULL);
+	assert(uri->pathHead != NULL);
+	assert(uri->pathHead->text.first == uri->pathHead->text.afterLast);
+
+	{
+		URI_TYPE(PathSegment) * const originalHead = uri->pathHead;
+
+		uri->pathHead = uri->pathHead->next;
+
+		originalHead->text.first = NULL;
+		originalHead->text.afterLast = NULL;
+		memory->free(memory, originalHead);
+	}
+}
+
+
+
+static void URI_FUNC(TransformEmptyFirstPathSegment)(URI_TYPE(Uri) * uri, UriMemoryManager * memory) {
+	assert(uri != NULL);
+	assert(memory != NULL);
+
+	if ((URI_FUNC(HasHost)(uri) == URI_TRUE)
+			|| (uri->pathHead == NULL)
+			|| (uri->pathHead->text.first != uri->pathHead->text.afterLast)) {
+		return;  /* i.e. nothing to do */
+	}
+
+	assert(uri->absolutePath == URI_FALSE);
+
+	URI_FUNC(DropEmptyFirstPathSegment)(uri, memory);
+
+	uri->absolutePath = URI_TRUE;
+
+	if ((uri->pathHead == uri->pathTail) && (uri->pathHead->text.first == uri->pathHead->text.afterLast)) {
+		URI_FUNC(DropEmptyFirstPathSegment)(uri, memory);
+	}
+}
+
+
+
+static int URI_FUNC(InternalSetPath)(URI_TYPE(Uri) * destUri,
+		const URI_CHAR * first,
+		const URI_CHAR * afterLast,
+		UriMemoryManager * memory) {
+	assert(destUri != NULL);
+	assert(first != NULL);
+	assert(afterLast != NULL);
+	assert(memory != NULL);
+	assert(destUri->pathHead == NULL);  /* set by SetPathMm right before */
+	assert(destUri->pathTail == NULL);  /* set by SetPathMm right before */
+	assert(destUri->absolutePath == URI_FALSE);  /* set by SetPathMm right before */
+
+	/* Skip the leading slash from target URIs with a host (so that we can
+	 * transfer the path 1:1 further down) */
+	if (URI_FUNC(HasHost)(destUri) == URI_TRUE) {
+		/* NOTE: This is because SetPathMm called IsWellFormedPath earlier: */
+		assert((afterLast - first >= 1) && (first[0] == _UT('/')));
+		first++;
+	} else if (first == afterLast) {
+	    /* This avoids (1) all the expensive but unecessary work below
+	     * and also (2) mis-encoding as single empty path segment
+	     * that would need (detection and) repair further down oetherwise */
```

Comment:
```suggestion
	     * that would need (detection and) repair further down otherwise */
```

Reply:
@TimWolla good catch. I should add codespell to CI here some time also, happily using that elsewhere.

Reply:
This thread is resolved.